### ISSUE TITLE
Simulate IO Cancellation for libaio

### ISSUE THREAD

**Issue by chamarthy** [axboe/fio#292]
Is there any way in FIO to simulate IO cancellations? I see that the aio engines have implementation of io_cancel in the code, but do not know a way to explicitly do cancellations for pending IOs.

# grep io_cancel * -r
engines/libaio.c:static int fio_libaio_cancel(struct thread_data *td, struct io_u *io_u)
engines/libaio.c:       return io_cancel(ld->aio_ctx, &io_u->iocb, ld->aio_events);
engines/libaio.c:       .cancel                 = fio_libaio_cancel,
engines/solarisaio.c:static int fio_solarisaio_cancel(struct thread_data fio_unused *td,
engines/solarisaio.c:   .cancel         = fio_solarisaio_cancel,
Binary file engines/libaio.o matches
Binary file engines/posixaio.o matches
engines/posixaio.c:static int fio_posixaio_cancel(struct thread_data fio_unused *td,
engines/posixaio.c:     int r = aio_cancel(f->fd, &io_u->aiocb);
engines/posixaio.c:     .cancel         = fio_posixaio_cancel,

```
370 static struct ioengine_ops ioengine = {
371   .name     = "libaio",
372   .version    = FIO_IOOPS_VERSION,
373   .init     = fio_libaio_init,
374   .prep     = fio_libaio_prep,
375   .queue      = fio_libaio_queue,
376   .commit     = fio_libaio_commit,
377   .cancel     = fio_libaio_cancel,
378   .getevents    = fio_libaio_getevents,
379   .event      = fio_libaio_event,
380   .cleanup    = fio_libaio_cleanup,
381   .open_file    = generic_open_file,
382   .close_file   = generic_close_file,
383   .get_file_size    = generic_get_file_size,
384   .options    = options,
385   .option_struct_size = sizeof(struct libaio_options),
386 };
```

Thanks for your support.

**Comment by axboe** [comment]
Fio only uses cancel when the job exits and we're doing cleanup. Arguably it should just be removed, as the in-kernel support (on Linux, at least) for IO cancellation is poor. That leaves solaris and the posixaio variant. Not sure it's worth supporting at all.

**Comment by chamarthy** [comment]
Thanks for the info Jens. We have a requirement to simulate IO cancellations in flight and validate the functionality and was looking at options in FIO.

**Comment by axboe** [comment]
What IO engine and what storage/OS stack?

It would not be hard to make that work in fio, since we're tracking the requests on the software side anyway. Issuing an occasional cancel based on some option defined criteria would be trivial.

**Comment by chamarthy** [comment]
We are using libaio / windowsaio / posixaio as of now. Storage stack is EMC Midrange VNXe. We plan to issue abort/cancel from Linux/Windows hosts using the above mentioned engines. As you said solarisaio / posixaio has somewhat better implementation, we can move to that too.

It would be great if there is a option for example libaio engine with libaio_cancel option that will take a % as parameter and do cancellations or whichever you deem good enough.

Thanks for your support.

**Comment by craigster0** [comment]
i'm not an fio developer or contributor, but i have to wonder, why do you
care about this?  what semantics do you except for i/o cancellation?

here's a couple of things to consider:

   - if you request an i/o, and then cancel it, there's no guarantee as to
   whether the i/o completed or was cancelled.  in fact, if the i/o request is
   large, or it involves multiple i/o locations, then cancelling it might
   result in part of the i/o request completing and part of it not (just as if
   the host had rebooted while the i/o request was in progress)
   - with the possible exception of POSIX aio (as implemented by GNU libc),
   the window to successfully cancel asynchronous i/o requests may be small.
   once they get sent to the disk driver (which will typically be quickly),
   they probably can't be cancelled,, which means there's effectively no
   ability to cancel i/o requests

--craig

**Comment by sitsofe** [comment]
@craigster0 Could it be that more than cancellation what @chamarthy wants is the ability to repeatedly send aborts if an IO engine would directly support it?

**Comment by chamarthy** [comment]
@sitsofe Yes. That's what exactly i was looking at. We want to test the SAN array to verify how the components behave when aborts are received.

**Comment by sitsofe** [comment]
OK that's understandable but it's still not straightforward. Things to note (mostly already pointed out by @craigster0 )
* It's not easy to abort I/O because the window for actually aborting can be small - you've got to catch the I/O while it is in the target device's queue but before the target device has completed it.
* Aborting will work best with I/O engines that are "closer" to the disk as there are less layers that will see the abort and just do a cancel on their local queue before it is sent.
* I'm not sure you can know what state the data is in at a "successfully" aborted LBA. Perhaps it's the old state, perhaps it's the new, perhaps it's something in-between or none of those at all...
* At a lower level aborting different commands can exercise different paths. For example in SCSI it turns out there are more potential abort interruption points for COMPARE AND WRITE compared to a plain WRITE...

What I would envision is an I/O "direction" of abortread or abortwrite that initially sends normal I/O. After the I/O is somehow marked as in-flight (so it's at least no longer queued within fio itself) but before it comes back as completed you would then try and send an abort for the in-flight I/O. If you wanted to be fancy you could try and count how many such aborts came back as successful because that would give a vague indication of proper aborts. I don't know how well this could be made to work when doing network file based I/O (does NFS have any concept of abort?) so it's a niche mode. You can probably tell I am thinking of it in terms of SCSI commands.
